src/credentials/GroupDataProviderImpl.cpp

```diff
@@ -0,0 +1,1312 @@
+        {
+            // New fabric, or new endpoint
+            GroupMapStore(fabric_index, mapping.endpoint, mapping.group).Save(GroupInfo(mapping.name));
+            EndpointStore(fabric_index, mapping.endpoint).Save(EndpointInfo(mapping.group));
+            fabric_info.first_endpoint = mapping.endpoint;
+            fabric_info.endpoint_count = 1;
+            return fabric_store.Save(fabric_info);
+        }
+
+        // Existing fabric
+
+        chip::EndpointId endpoint_id   = fabric_info.first_endpoint;
+        chip::EndpointId prev_endpoint = kInvalidEndpointId;
+        EndpointInfo endpoint_info;
+        EndpointInfo prev_end_info;
+        size_t endpoint_count = 0;
+
+        // Loop through the fabric's endpoints
+        do
+        {
+            EndpointStore endpoint_store(fabric_index, endpoint_id);
+            if (CHIP_NO_ERROR != endpoint_store.Load(endpoint_info))
+            {
+                // Endpoint info not found
+                break;
+            }
+            if (endpoint_id == mapping.endpoint)
+            {
+                // Target endpoint info found
+                break;
+            }
+            prev_end_info = endpoint_info;
+            prev_endpoint = endpoint_id;
+            endpoint_id   = endpoint_info.next;
+        } while (++endpoint_count < fabric_info.endpoint_count);
+
+        if (endpoint_id != mapping.endpoint)
+        {
+            // Endpoint info NOT found, create new
+            if (kInvalidEndpointId != prev_endpoint)
+            {
+                // Link to existing endpoint
+                prev_end_info.next = mapping.endpoint;
+                EndpointStore(fabric_index, prev_endpoint).Save(prev_end_info);
+            }
+            // Append new endpoint info
+            GroupMapStore(fabric_index, mapping.endpoint, mapping.group).Save(GroupInfo(mapping.name));
+            EndpointStore(fabric_index, mapping.endpoint).Save(EndpointInfo(mapping.group));
+            fabric_info.endpoint_count++;
+            return fabric_store.Save(fabric_info);
+        }
+
+        // Endpoint info found, loop endpoints
+
+        chip::GroupId group_id   = endpoint_info.first_group;
+        chip::GroupId prev_group = kUndefinedGroupId;
+        GroupInfo group_info;
+
+        do
+        {
+            if (CHIP_NO_ERROR != GroupMapStore(fabric_index, endpoint_id, group_id).Load(group_info))
+            {
+                break;
+            }
+            if (group_id == mapping.group)
+            {
+                // Duplicated group
+                return CHIP_NO_ERROR;
+            }
+            prev_group = group_id;
+            group_id   = group_info.next;
+        } while ((kUndefinedGroupId != group_id) && (prev_group != group_id));
+
+        GroupMapStore(fabric_index, mapping.endpoint, mapping.group).Save(GroupInfo(mapping.name));
+        if (kUndefinedGroupId == prev_group)
```

Comment:
Again, if we inserted at front of list we could probably simplify the logic here some.

Does something in the spec require us to append?  I don't think append vs insert is observable on that level....

Reply:
Correct. The new entries are now inserted first.